### PR TITLE
At/#2144 error boundaries

### PR DESCRIPTION
**Pull request by AlexeyTuboltsev**
-create an ErrorBoundary component
-create a HOC (createErrorBoundary) that takes a component to display error
-this HOC has the same signature as all recompose functions, so we can use it in a branch() etc

-split Panel into PanelBase, Header and Panel components for better composability

-remove some unused stuff

### CODE REVIEW

File: app/components/shared/ErrorBoundary.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+interface IErrorBoundaryState {
+  hasError: boolean;
+}
+
+class ErrorBoundary extends React.Component<any, IErrorBoundaryState> {
+  state = { hasError: false };
+
+  componentDidCatch(error: Error | null): void {
+    captureException(error); //Error goes to Sentry
```
Comment (desfero) [inline app/components/shared/ErrorBoundary.tsx]: Can we for consistency use our `Logger` here and log `fatal` error?

File: app/components/shared/ErrorBoundary.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import { ComponentEnhancer, nest, withProps } from "recompose";
+
+interface IErrorBoundaryState {
+  hasError: boolean;
+}
+
+class ErrorBoundary extends React.Component<any, IErrorBoundaryState> {
+  state = { hasError: false };
+
+  componentDidCatch(error: Error | null): void {
```
Comment (desfero) [inline app/components/shared/ErrorBoundary.tsx]: please log also `errorInfo` like in `App.tsx`

File: app/components/shared/ErrorBoundary.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  render(): React.ReactNode {
+    if (this.state.hasError) {
+      return <this.props.layout />;
+    }
+
+    return this.props.children;
+  }
+}
+
+const createErrorBoundary = (layout: any): ComponentEnhancer<any, any> => {
```
Comment (desfero) [inline app/components/shared/ErrorBoundary.tsx]: `layout: React.ReactNode`

File: app/components/shared/ErrorBoundary.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { captureException } from "@sentry/browser";
+import * as React from "react";
+import { ComponentEnhancer, nest, withProps } from "recompose";
+
+interface IErrorBoundaryState {
+  hasError: boolean;
+}
+
+class ErrorBoundary extends React.Component<any, IErrorBoundaryState> {
```
Comment (desfero) [inline app/components/shared/ErrorBoundary.tsx]: can we add this to visual regression?

File: app/components/shared/EtoFormProgressWidget.tsx [hunk 5]
```diff
@@ -1,13 +1,16 @@
 import * as React from "react";
 import { FormattedMessage } from "react-intl-phraseapp";
 
+import { compose } from "redux";
```
Comment (desfero) [inline app/components/shared/EtoFormProgressWidget.tsx]: should be in a group above

File: app/components/shared/EtoFormProgressWidget.tsx [hunk 6]
```diff
@@ -1,13 +1,16 @@
...
+import * as arrowRightIcon from "../../assets/img/inline_icons/arrow_right.svg";
```
Comment (desfero) [inline app/components/shared/EtoFormProgressWidget.tsx]: should be in a group below

File: app/components/shared/EtoFormProgressWidget.tsx [hunk 7]
```diff
@@ -34,7 +37,7 @@ const ButtonText: React.SFC<IButtonTextProps> = ({ isInProgress, readonly }) =>
   return <FormattedMessage id="shared-component.eto-form-progress-widget.edit" />;
 };
 
-export const EtoFormProgressWidget: React.SFC<IProps & IChartCircleProps> = ({
+export const EtoFormProgressWidgetComponent: React.SFC<IProps & IChartCircleProps> = ({
```
Comment (desfero) [inline app/components/shared/EtoFormProgressWidget.tsx]: We agree to follow convetion `EtoFormProgressWidgetLayout` for layout components

File: app/components/shared/ErrorBoundary.tsx (hunk 2 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/ErrorBoundary.tsx]: the problem here is that it's a huge pain in the ass to get this logger, cause i need to setup all the context

File: app/components/shared/ErrorBoundary.tsx (hunk 1 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/ErrorBoundary.tsx]: the problem here is that it's a huge pain in the ass to get this logger, 'cause I need to setup the entire context for this component. Not sure if it's worth it

File: app/components/shared/ErrorBoundary.tsx (hunk 3 above)
Comment (AlexeyTuboltsev) [reply app/components/shared/ErrorBoundary.tsx]: ❤️

File: app/components/shared/Panel.tsx [hunk 8]
```diff
@@ -26,18 +28,14 @@ const Panel: React.SFC<IPanelProps> = ({
...
   return (
-    <div {...props} className={cn(styles.panel, className, { [styles.narrow]: narrow })}>
+    <PanelBase {...props} className={className} narrow={narrow}>
       {hasHeader && (
-        <header className={cn(styles.header, { [styles.hasIcon]: !!(icon || !headerText) })}>
-          {icon && <img src={icon} className={styles.icon} />}
-          {headerText && <div className={styles.left}>{headerText}</div>}
-          {rightComponent && <div className={styles.right}>{rightComponent}</div>}
-        </header>
+        <PanelHeader icon={icon} headerText={headerText} rightComponent={rightComponent} />
```
[the remark this answers is not in the document]
Comment (AlexeyTuboltsev) [reply app/components/shared/Panel.tsx]: well, no...

**Pull request merged**